crypto/o_str.c

```diff
@@ -338,3 +346,95 @@ int openssl_strerror_r(int errnum, char *buf, size_t buflen)
     return 1;
 #endif
 }
+
+#ifndef OPENSSL_NO_LOCALE
+# ifndef FIPS_MODULE
+static CRYPTO_ONCE casecmp = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(init_casecmp)
+{
+    int ret = ossl_init_casecmp_int();
+
+    return ret;
+}
+
+int ossl_init_casecmp(void)
+{
+    if (!RUN_ONCE(&casecmp, init_casecmp))
+        return 0;
+    return 1;
+}
+# endif
+
+static locale_t loc;
+
+static locale_t ossl_c_locale(void)
+{
+# ifndef FIPS_MODULE
+    if (!ossl_init_casecmp())
+        return (locale_t)0;
+# endif
+    return loc;
+}
+
+int ossl_init_casecmp_int(void)
+{
+# ifdef OPENSSL_SYS_WINDOWS
+    loc = _create_locale(LC_COLLATE, "C");
+# else
+    loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
+# endif
+    return (loc == (locale_t)0) ? 0 : 1;
+}
+
+void ossl_deinit_casecmp(void)
+{
+    if (loc != (locale_t)0) {
+        freelocale(loc);
+        loc = (locale_t)0;
+    }
+}
+
+int OPENSSL_strcasecmp(const char *s1, const char *s2)
+{
+    locale_t l = ossl_c_locale();
+
+    /* Fallback in case of locale initialization failure */
+    if (l == (locale_t)0)
+        return strcasecmp(s1, s2);
+    return strcasecmp_l(s1, s2, l);
```

Comment:
This fallback masks the cases when we lack the initialization, doesn't it?

Reply:
IIRC, `loc` is required to initialize to `0` by C standards whether in a DLL or program based on it being on the initial global heap.

Reply:
No, it comes into an effect only if the initialization fails. The last commit ensures that we always run the RUN_ONCE from OPENSSL_strcasecmp as we discussed in OTC.

Reply:
Sorry, my fault